// @trimble-oss/modus-web-components/dist/collection/components/modus-table/parts/panel/modus-table-toolbar/modus-table-toolbar.js
import { Host, h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from "@stencil/core";
export class ModusTablePanel {
    constructor() {
        this.context = undefined;
    }
    render() {
        // const { tableInstance: table, toolbarOptions: options } = this.context;
        return (h(Host, { key: 'd8b9a5275dc4e77bdacb4958d1cdd4079872302c' }, h("div", { key: '5f5d87db6a9cd8dd774e055396fd12679d8c534e', class: "table-toolbar" }, h("div", { key: '432d5965c73c369d63e55f81ce300dc9c52caee6', class: "section" }, h("slot", { key: '11786f71e2d0a2ac80bbf56a5e3de148a57de865', name: "group-left" })), h("div", { key: '2b515ed9f8869b8c16ef478deb87fb5f629182cf', class: "section" }, h("slot", { key: 'ea1db6297b8534aa58ed9c5a5793b810ff7b8dc3', name: "group-right" }), h("modus-table-dropdown-menu", { context: this.context })))));
    }
    static get is() { return "modus-table-toolbar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["./modus-table-toolbar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-table-toolbar.css"]
        };
    }
    static get properties() {
        return {
            "context": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TableContext",
                    "resolved": "TableContext",
                    "references": {
                        "TableContext": {
                            "location": "import",
                            "path": "../../../models/table-context.models",
                            "id": "src/components/modus-table/models/table-context.models.ts::TableContext"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Table data."
                }
            }
        };
    }
}
