// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/button-menu/modus-navbar-button-menu.js
import { h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from "@stencil/core";
export class ModusNavbarButtonMenu {
    constructor() {
        this.reverse = undefined;
    }
    render() {
        const menuClasses = {
            'button-menu': true,
            reverse: this.reverse,
        };
        return (h("div", { key: 'e8656ea27ea109cc8149411ce9ecb2d1a0fc60f0', class: menuClasses }, h("slot", { key: '121ef3a22ddd0ff78ddb482b86c416d12405b83d' })));
    }
    static get is() { return "modus-navbar-button-menu"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-navbar-button-menu.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-navbar-button-menu.css"]
        };
    }
    static get properties() {
        return {
            "reverse": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "reverse",
                "reflect": false
            }
        };
    }
}
