// @trimble-oss/modus-web-components/dist/collection/components/modus-utility-panel/modus-utility-panel.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { h, Fragment } from "@stencil/core";
export class ModusUtilityPanel {
    constructor() {
        this.handlePanelClose = () => {
            this.closePanel();
        };
        this.expanded = false;
        this.pushContent = true;
        this.targetContent = undefined;
    }
    componentDidLoad() {
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    handleExpandedChange(newValue) {
        if (newValue) {
            this.openPanel();
        }
        else {
            this.closePanel();
        }
    }
    async openPanel() {
        this.panelOpened.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    async closePanel() {
        this.panelClosed.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    adjustContent() {
        const content = document.querySelector(this.targetContent);
        if (content) {
            content.style.transition = 'margin-right 0.2s ease-out';
            if (this.expanded) {
                content.style.marginRight = '312px';
            }
            else {
                content.style.marginRight = '0';
            }
        }
    }
    hasSlotContent(slotName) {
        const slot = this.el.querySelector(`[slot="${slotName}"]`);
        return !!slot;
    }
    render() {
        const hasHeader = this.hasSlotContent('header');
        const hasFooter = this.hasSlotContent('footer');
        return (h("div", { key: 'fa7fff3598fff4a3b5a1fc33e8d798dc432e63d0', class: { 'utility-panel': true, open: this.expanded, overlay: !this.pushContent } }, h("div", { key: '0fda6d981bce0f9d78f46e0763c1d394a3575a95', class: "panel-content" }, hasHeader && (h(Fragment, null, h("div", { class: "panel-header", "aria-labelledby": "header" }, h("slot", { name: "header" })), h("hr", null))), h("div", { key: 'f4a4629cb9bbf518df1e1b3c6adc13e62ba07334', class: "panel-body", "aria-labelledby": "body" }, h("slot", { key: 'acb99a4c2d5f3858b8d324b0172bb96fb7c38743', name: "body" })), hasFooter && (h(Fragment, null, h("hr", null), h("div", { class: "panel-footer", "aria-labelledby": "footer" }, h("slot", { name: "footer" })))))));
    }
    static get is() { return "modus-utility-panel"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-utility-panel.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-utility-panel.css"]
        };
    }
    static get properties() {
        return {
            "expanded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The panel is expanded or closed"
                },
                "attribute": "expanded",
                "reflect": false,
                "defaultValue": "false"
            },
            "pushContent": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines if the panel pushes content or displays an overlay."
                },
                "attribute": "push-content",
                "reflect": false,
                "defaultValue": "true"
            },
            "targetContent": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "target-content",
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "panelOpened",
                "name": "panelOpened",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires when the panel is opened."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "panelClosed",
                "name": "panelClosed",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires when the panel is closed."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "expanded",
                "methodName": "handleExpandedChange"
            }];
    }
}
