// @trimble-oss/modus-web-components/dist/collection/components/modus-table/parts/fillerColumn/modus-table-filler-column.js
import { h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from "@stencil/core";
/**
 * ModusFillerColumn is to fill empty space within a table or grid when the content in other columns is not wide enough to occupy the entire available width
 */
export class ModusTableFillerColumn {
    constructor() {
        this.observer = null;
        this.updateContainerLayout = () => {
            var _a, _b, _c, _d, _e, _f;
            const tableWidth = (_a = this.targetTable.getBoundingClientRect()) === null || _a === void 0 ? void 0 : _a.width;
            const parentWidth = (_c = (_b = this.targetTable.parentElement) === null || _b === void 0 ? void 0 : _b.getBoundingClientRect()) === null || _c === void 0 ? void 0 : _c.width;
            this.showFillerTable = tableWidth < parentWidth;
            if (!this.showFillerTable)
                return;
            if (this.fillerTableRef) {
                this.fillerTableRef.querySelector('thead').style.height = `${(_d = this.targetTable.querySelector('thead')) === null || _d === void 0 ? void 0 : _d.getBoundingClientRect().height}px`;
                if (this.summaryRow) {
                    this.fillerTableRef.querySelector('tfoot').style.height = `${(_e = this.targetTable.querySelector('tfoot')) === null || _e === void 0 ? void 0 : _e.getBoundingClientRect().height}px`;
                }
                this.fillerTableRef.querySelector('tbody').style.height = `${(_f = this.targetTable.querySelector('tbody')) === null || _f === void 0 ? void 0 : _f.getBoundingClientRect().height}px`;
            }
        };
        this.cellBorderless = undefined;
        this.summaryRow = undefined;
        this.container = undefined;
        this.showFillerTable = false;
    }
    componentDidLoad() {
        this.targetTable = this.container.shadowRoot.querySelector('table');
        if (this.targetTable) {
            this.updateContainerLayout();
            this.connectDOMObserver();
        }
    }
    disconnectedCallback() {
        this.disconnectDOMObserver();
    }
    connectDOMObserver() {
        this.observer = new ResizeObserver(this.updateContainerLayout);
        this.observer.observe(this.targetTable);
    }
    disconnectDOMObserver() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    render() {
        return (h("table", { key: '8c5d31136767a6da879b04edca26d4da23e667a5', id: "table-filler-column", class: { 'cell-borderless': this.cellBorderless, 'd-none': !this.showFillerTable }, ref: (el) => (this.fillerTableRef = el) }, h("thead", { key: '458db94209103de571922b9b0bb05146c8033f45' }, h("tr", { key: '560d80b509a9bc83f851e972d7d77f68bebfa085' }, h("th", { key: '760ed56de2d8d909e5e2e17b25b56148a85537ae' }))), h("tbody", { key: 'fc01fb5caf8078b4989268ddb2b56a27820e3d06' }, h("tr", { key: '1dda921e8ce27b3538ef858829c5215329319122' }, h("td", { key: '2b9208f58aa2f45dac97186d33433ce4e450382e' }))), this.summaryRow && (h("tfoot", null, h("tr", { class: "summary-row" }, h("td", null))))));
    }
    static get is() { return "modus-table-filler-column"; }
    static get originalStyleUrls() {
        return {
            "$": ["./modus-table-filler-column.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-table-filler-column.css"]
        };
    }
    static get properties() {
        return {
            "cellBorderless": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "cell-borderless",
                "reflect": false
            },
            "summaryRow": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "summary-row",
                "reflect": false
            },
            "container": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "HTMLElement",
                    "resolved": "HTMLElement",
                    "references": {
                        "HTMLElement": {
                            "location": "global",
                            "id": "global::HTMLElement"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            }
        };
    }
    static get states() {
        return {
            "showFillerTable": {}
        };
    }
}
