// @trimble-oss/modus-web-components/dist/cjs/modus-toast.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-8ad089ae.js');
const iconError = require('./icon-error-98669c7d.js');
const iconWarning = require('./icon-warning-f22142a3.js');
const iconHelp = require('./icon-help-b14f59ac.js');
const IconClose = require('./IconClose-cec27da2.js');

const modusToastCss = ".modus-toast{align-items:center;background-color:var(--modus-toast-bg, #fff);border:0.0625rem solid transparent;border-radius:0.25rem;box-shadow:0 0 8px rgba(36, 35, 45, 0.3);color:var(--modus-toast-color, #252a2e);display:flex;flex-direction:row;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";font-size:0.875rem;min-height:2.6875rem;padding:0 1rem;width:370px}.modus-toast .icon{align-items:center;display:flex;justify-content:center;min-width:1.125rem}.modus-toast .text{margin:0 0.5rem}.modus-toast .close{background-color:transparent;border:0;margin-left:auto;margin-top:2px}.modus-toast .close .icon-close{height:18px;width:18px}.modus-toast .close:hover svg{cursor:pointer}.modus-toast .close:hover svg path{opacity:0.5}.modus-toast.primary{background-color:var(--modus-toast-primary-bg, #0063a3);border-color:var(--modus-toast-primary-bg, );color:var(--modus-toast-primary-color, #fff)}.modus-toast.primary .close svg path{fill:var(--modus-toast-primary-color, var(--modus-toast-color, #252a2e))}.modus-toast.primary svg:not(.icon-close) path{fill:var(--modus-toast-primary-color, )}.modus-toast.secondary{background-color:var(--modus-toast-secondary-bg, #252a2e);border-color:var(--modus-toast-secondary-bg, );color:var(--modus-toast-secondary-color, #fff)}.modus-toast.secondary .close svg path{fill:var(--modus-toast-secondary-color, var(--modus-toast-color, #252a2e))}.modus-toast.secondary svg:not(.icon-close) path{fill:var(--modus-toast-secondary-color, )}.modus-toast.success{background-color:var(--modus-toast-success-bg, #006638);border-color:var(--modus-toast-success-bg, );color:var(--modus-toast-success-color, #fff)}.modus-toast.success .close svg path{fill:var(--modus-toast-success-color, var(--modus-toast-color, #252a2e))}.modus-toast.success svg:not(.icon-close) path{fill:var(--modus-toast-success-color, )}.modus-toast.danger{background-color:var(--modus-toast-danger-bg, #ab1f26);border-color:var(--modus-toast-danger-bg, );color:var(--modus-toast-danger-color, #fff)}.modus-toast.danger .close svg path{fill:var(--modus-toast-danger-color, var(--modus-toast-color, #252a2e))}.modus-toast.danger svg:not(.icon-close) path{fill:var(--modus-toast-danger-color, )}";
const ModusToastStyle0 = modusToastCss;

const ModusToast = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dismissClick = index.createEvent(this, "dismissClick", 7);
        this.iconByType = new Map([
            ['danger', index.h(iconWarning.IconWarning, { size: '18' })],
            ['dark', index.h(iconWarning.IconInfo, { size: '18' })],
            ['default', index.h(iconWarning.IconInfo, { size: '18' })],
            ['primary', index.h(iconWarning.IconInfo, { size: '18' })],
            ['secondary', index.h(iconHelp.IconHelp, { size: '18' })],
            ['success', index.h(iconWarning.IconCheckCircle, { size: '18' })],
            ['tertiary', index.h(iconWarning.IconInfo, { size: '18' })],
            ['warning', index.h(iconError.IconError, { size: '18' })],
        ]);
        this.classByType = new Map([
            ['danger', 'danger'],
            ['dark', 'dark'],
            ['default', 'default'],
            ['primary', 'primary'],
            ['secondary', 'secondary'],
            ['success', 'success'],
            ['tertiary', 'tertiary'],
            ['warning', 'warning'],
        ]);
        this.ariaLabel = undefined;
        this.dismissible = undefined;
        this.delay = 15000;
        this.retainElement = false;
        this.role = 'status';
        this.showIcon = true;
        this.type = 'default';
    }
    delayChanged(newDelay) {
        clearTimeout(this.timerId);
        this.timerId = setTimeout(() => {
            this.dismissElement();
        }, newDelay);
    }
    dismissElement() {
        this.dismissClick.emit();
        if (!this.retainElement) {
            this.el.remove();
        }
    }
    componentDidLoad() {
        if (this.delay > 0) {
            this.timerId = setTimeout(() => {
                this.dismissElement();
            }, this.delay);
        }
    }
    disconnectedCallback() {
        clearTimeout(this.timerId);
    }
    render() {
        const icon = this.iconByType.get(this.type);
        const className = `modus-toast ${this.classByType.get(this.type)}`;
        return (index.h("div", { key: '8658f1fee085330df445f03ac14d44573ce4de6c', "aria-label": this.ariaLabel || undefined, class: className, role: this.role }, this.showIcon && index.h("div", { class: "icon" }, icon), index.h("span", { key: 'e4751abc2649fd769044abf05cee339da6eba12b', class: 'text' }, index.h("slot", { key: '234833e6f3412f94e1c1b7ea2a5d3e5c1346cc2b' })), this.dismissible && (index.h("button", { type: "button", class: 'close', onClick: () => this.dismissElement(), "aria-label": "Dismiss" }, index.h(IconClose.IconClose, { size: '18' })))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "delay": ["delayChanged"]
    }; }
};
ModusToast.style = ModusToastStyle0;

exports.modus_toast = ModusToast;
