// @trimble-oss/modus-web-components/dist/cjs/modus-utility-panel.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-8ad089ae.js');

const modusUtilityPanelCss = ":host{--panel-padding:1rem}.utility-panel{background-color:var(--modus-utility-panel-bg, #fff);box-shadow:0 0 10px rgba(0, 0, 0, 0.5);color:var(--modus-utility-panel-color, #000);height:100%;position:absolute;right:-324px;top:0;transition:right 0.3s ease-out;width:312px}.utility-panel.open{right:0;transition:right 0.3s ease-out}.utility-panel .overlay{height:100%;position:absolute;right:0;top:0;width:100%}.utility-panel .panel-content{display:flex;flex-direction:column;height:100%}.utility-panel .panel-content .panel-header,.utility-panel .panel-content .panel-footer{align-items:center;display:flex;height:50px;padding:0 var(--panel-padding)}.utility-panel .panel-content .panel-body{flex:1;overflow:auto;padding:var(--panel-padding)}.utility-panel .panel-content hr{border:none;border-top:1px solid #cbcdd6;margin:0}";
const ModusUtilityPanelStyle0 = modusUtilityPanelCss;

const ModusUtilityPanel = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.panelOpened = index.createEvent(this, "panelOpened", 7);
        this.panelClosed = index.createEvent(this, "panelClosed", 7);
        this.handlePanelClose = () => {
            this.closePanel();
        };
        this.expanded = false;
        this.pushContent = true;
        this.targetContent = undefined;
    }
    componentDidLoad() {
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    handleExpandedChange(newValue) {
        if (newValue) {
            this.openPanel();
        }
        else {
            this.closePanel();
        }
    }
    async openPanel() {
        this.panelOpened.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    async closePanel() {
        this.panelClosed.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    adjustContent() {
        const content = document.querySelector(this.targetContent);
        if (content) {
            content.style.transition = 'margin-right 0.2s ease-out';
            if (this.expanded) {
                content.style.marginRight = '312px';
            }
            else {
                content.style.marginRight = '0';
            }
        }
    }
    hasSlotContent(slotName) {
        const slot = this.el.querySelector(`[slot="${slotName}"]`);
        return !!slot;
    }
    render() {
        const hasHeader = this.hasSlotContent('header');
        const hasFooter = this.hasSlotContent('footer');
        return (index.h("div", { key: 'fa7fff3598fff4a3b5a1fc33e8d798dc432e63d0', class: { 'utility-panel': true, open: this.expanded, overlay: !this.pushContent } }, index.h("div", { key: '0fda6d981bce0f9d78f46e0763c1d394a3575a95', class: "panel-content" }, hasHeader && (index.h(index.Fragment, null, index.h("div", { class: "panel-header", "aria-labelledby": "header" }, index.h("slot", { name: "header" })), index.h("hr", null))), index.h("div", { key: 'f4a4629cb9bbf518df1e1b3c6adc13e62ba07334', class: "panel-body", "aria-labelledby": "body" }, index.h("slot", { key: 'acb99a4c2d5f3858b8d324b0172bb96fb7c38743', name: "body" })), hasFooter && (index.h(index.Fragment, null, index.h("hr", null), index.h("div", { class: "panel-footer", "aria-labelledby": "footer" }, index.h("slot", { name: "footer" })))))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "expanded": ["handleExpandedChange"]
    }; }
};
ModusUtilityPanel.style = ModusUtilityPanelStyle0;

exports.modus_utility_panel = ModusUtilityPanel;
