// @trimble-oss/modus-web-components/dist/cjs/modus-list_2.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-8ad089ae.js');
const iconCheck = require('./icon-check-a22357b3.js');
const ModusIconMap = require('./ModusIconMap-4076f8df.js');
require('./IconChevronLeft-cd7905fc.js');
require('./IconChevronRight-3b379c08.js');
require('./IconClose-cec27da2.js');
require('./IconExpand-15ffe311.js');
require('./IconVisibilityOn-0cf159be.js');

const modusListCss = "ul{background-color:var(--modus-list-bg, #fff);border-radius:0.25rem;display:flex;flex-direction:column;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";list-style:none;margin:0;padding:0;position:relative}";
const ModusListStyle0 = modusListCss;

const ModusList = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    handleKeyDown(e) {
        const itemsLength = this.element.children.length;
        if (e.key.toLowerCase() === 'arrowdown') {
            const index = Array.prototype.indexOf.call(this.element.children, e.target);
            let next = this.element.children.item((index + 1) % itemsLength);
            while (next === null || next === void 0 ? void 0 : next.disabled) {
                next = this.element.children.item((index + 2) % itemsLength);
            }
            next === null || next === void 0 ? void 0 : next.focusItem();
            e.preventDefault();
        }
        if (e.key.toLowerCase() === 'arrowup') {
            const index = Array.prototype.indexOf.call(this.element.children, e.target);
            let prev = this.element.children.item((index - 1) % itemsLength);
            while (prev === null || prev === void 0 ? void 0 : prev.disabled) {
                prev = this.element.children.item((index - 2) % itemsLength);
            }
            prev === null || prev === void 0 ? void 0 : prev.focusItem();
            e.preventDefault();
        }
    }
    render() {
        return (index.h("ul", { key: 'd50f4a349d16e3e76bca4e5923ffafb6580d9a80', part: "list-items", onKeyDown: (e) => this.handleKeyDown(e) }, index.h("slot", { key: '61d4f7de300ae52f88732311aa5c0b81ae3acbb8' })));
    }
    get element() { return index.getElement(this); }
};
ModusList.style = ModusListStyle0;

const modusListItemCss = "li{align-items:center;background-color:var(--modus-list-item-bg, #fff);border:1px solid var(--modus-list-item-border-color, #e0e1e9);border-radius:4px;box-sizing:border-box;color:var(--modus-list-item-color, #252a2e);display:flex;fill:var(--modus-list-item-color, #252a2e);font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";gap:1rem;justify-content:space-between;min-height:2.5rem;padding:0.375rem 0.75rem}li .text-container{display:flex;flex-direction:column;flex-grow:1;justify-content:center;min-width:0}li .slot{font-size:0.875rem;line-height:1.125rem;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}li .sub-text{display:block;font-size:0.75rem;line-height:1rem}li .wrap{word-wrap:break-word}li .no-wrap{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}li .icon-check{flex-shrink:0;margin-left:auto}li.borderless,li.borderless:hover:not(.disabled){border:none}li:hover:not(.disabled){background-color:var(--modus-list-item-hover-bg, #e0e1e9);border:1px solid var(--modus-list-item-border-color, #e0e1e9);color:var(--modus-list-item-color, #252a2e);fill:var(--modus-list-item-color, #252a2e)}li.small{gap:0.625rem;min-height:2rem;padding:0.25rem 0.625rem}li.small .slot{font-size:0.75rem;line-height:0.875rem}li.small .sub-text{font-size:0.625rem;line-height:0.75rem}li.large{min-height:3rem;padding:0.625rem 1rem}li.disabled{color:var(--modus-list-item-disabled-color, #b7b9c3);cursor:default;fill:var(--modus-list-item-disabled-color, #b7b9c3)}li.selected{background-color:var(--modus-list-item-selected-bg, #dcedf9);border:1px solid var(--modus-list-item-selected-border-color, #dcedf9);color:var(--modus-list-item-color, #252a2e);fill:var(--modus-list-item-color, #252a2e)}@media screen and (forced-colors: active){li.disabled{border-color:GrayText;color:GrayText}}";
const ModusListItemStyle0 = modusListItemCss;

const ModusListItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.itemClick = index.createEvent(this, "itemClick", 7);
        this.classBySize = new Map([
            ['condensed', 'small'],
            ['standard', 'standard'],
            ['large', 'large'],
        ]);
        this.borderless = undefined;
        this.disabled = undefined;
        this.selected = undefined;
        this.leftIcon = undefined;
        this.size = 'standard';
        this.subText = undefined;
        this.wrapSubText = true;
        this.type = 'standard';
        this.iconColor = undefined;
    }
    async focusItem() {
        var _a;
        (_a = this.listItemRef) === null || _a === void 0 ? void 0 : _a.focus();
    }
    handleKeydown(e) {
        if (e.key.toLowerCase() === 'enter' && !this.disabled) {
            this.itemClick.emit();
        }
    }
    render() {
        const containerClass = `${this.classBySize.get(this.size)} ${this.disabled ? 'disabled' : ''} ${this.selected ? 'selected' : ''} ${this.borderless ? 'borderless' : ''}`;
        const iconCheckSize = this.size === 'condensed' ? '16' : '24';
        return (index.h("li", { key: 'a143f4008c7667ab3a118f8e4f243fdf845c9a25', ref: (el) => (this.listItemRef = el), part: "list-item-li", class: containerClass, tabIndex: this.disabled ? -1 : 0, onClick: () => (!this.disabled ? this.itemClick.emit() : null), onKeyDown: (e) => this.handleKeydown(e) }, this.leftIcon && (index.h("span", { class: "icon left-icon" }, index.h(ModusIconMap.ModusIconMap, { icon: this.leftIcon, size: "24", color: this.iconColor }))), index.h("div", { key: '6caeadbfe730b540ff4b34d3c1991c5c64a95add', class: "text-container" }, index.h("span", { key: 'd0d695a03abafd70b5a240178921492c0523e58c', class: "slot" }, index.h("slot", { key: 'f3f949fbe697d45e4ae93429de1c6e753354ff39' })), this.subText && index.h("span", { class: 'sub-text ' + (this.wrapSubText ? 'wrap' : 'no-wrap') }, this.subText)), this.selected && index.h(iconCheck.IconCheck, { size: iconCheckSize })));
    }
};
ModusListItem.style = ModusListItemStyle0;

exports.modus_list = ModusList;
exports.modus_list_item = ModusListItem;
