// @trimble-oss/modus-web-components/dist/collection/components/modus-progress-bar/modus-progress-bar.js
// eslint-disable-next-line
import { h, Host } from "@stencil/core";
export class ModusProgressBar {
    constructor() {
        this.classBySize = new Map([
            ['default', 'default'],
            ['small', 'small'],
            ['compact', 'compact'],
        ]);
        this.ariaLabel = undefined;
        this.backgroundColor = undefined;
        this.color = undefined;
        this.maxValue = 100;
        this.minValue = 0;
        this.mode = 'determinate';
        this.size = 'default';
        this.text = undefined;
        this.textColor = undefined;
        this.value = 0;
    }
    getProgressStyle(percentage) {
        const progressStyle = {
            backgroundColor: this.color,
            color: this.textColor,
            width: this.mode === 'determinate' ? `${percentage}%` : '100%',
        };
        return progressStyle;
    }
    getProgressBarStyle() {
        const progressBarStyle = {
            backgroundColor: this.backgroundColor,
        };
        return progressBarStyle;
    }
    render() {
        const percentage = ((this.value - this.minValue) / (this.maxValue - this.minValue)) * 100;
        const progressBarBackgroundColorClass = this.backgroundColor ? '' : 'default-background-color';
        const progressColorClass = this.color ? '' : 'default-color';
        const progressTextColor = this.textColor ? '' : 'default-text-color';
        const progressBarClass = `
      modus-progress-bar
      ${progressBarBackgroundColorClass}
      ${this.classBySize.get(this.size)}
     `;
        const progressClass = `progress ${progressColorClass} ${progressTextColor} ${this.mode === 'indeterminate' ? 'indeterminate' : 'determinate'}`;
        return (h(Host, { key: '598b24ac61bf998feea59b9188bd4ea9c2fb9b29', "aria-label": this.ariaLabel, "aria-valuemax": this.mode === 'determinate' ? this.maxValue : null, "aria-valuemin": this.mode === 'determinate' ? this.minValue : null, "aria-valuenow": this.mode === 'determinate' ? this.value : null, role: "progressbar" }, h("div", { key: '0c881a3a711ee478d624c11834c858aeef646348', class: progressBarClass, style: this.getProgressBarStyle() }, h("div", { key: 'd6f1f979aa8bdb2ddb63188b1052655e7392c072', class: progressClass, style: this.getProgressStyle(percentage) }, this.size === 'default' && this.mode === 'determinate' && this.text))));
    }
    static get is() { return "modus-progress-bar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-progress-bar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-progress-bar.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's aria-label."
                },
                "attribute": "aria-label",
                "reflect": false
            },
            "backgroundColor": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's background color."
                },
                "attribute": "background-color",
                "reflect": false
            },
            "color": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's foreground color."
                },
                "attribute": "color",
                "reflect": false
            },
            "maxValue": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's maximum value."
                },
                "attribute": "max-value",
                "reflect": false,
                "defaultValue": "100"
            },
            "minValue": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's minimum value."
                },
                "attribute": "min-value",
                "reflect": false,
                "defaultValue": "0"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'determinate' | 'indeterminate'",
                    "resolved": "\"determinate\" | \"indeterminate\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's mode."
                },
                "attribute": "mode",
                "reflect": false,
                "defaultValue": "'determinate'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'small' | 'compact'",
                    "resolved": "\"compact\" | \"default\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's size."
                },
                "attribute": "size",
                "reflect": false,
                "defaultValue": "'default'"
            },
            "text": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The text displayed on the progress bar."
                },
                "attribute": "text",
                "reflect": false
            },
            "textColor": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's text color."
                },
                "attribute": "text-color",
                "reflect": false
            },
            "value": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The progress bar's value."
                },
                "attribute": "value",
                "reflect": false,
                "defaultValue": "0"
            }
        };
    }
}
