// @trimble-oss/modus-web-components/dist/components/modus-card.js
import { p as proxyCustomElement, H, h } from './p-85681586.js';

const modusCardCss = "article{background-color:var(--modus-card-bg, #fff);border:0.0625rem solid;border-color:transparent;border-radius:0.25rem;color:var(--modus-card-color, #252a2e);font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";overflow:hidden;position:relative}article.shadow:hover{box-shadow:0 0 4px var(--modus-card-shadow-color, rgba(54, 53, 69, 0.3019607843))}article.card-border{border-color:var(--modus-card-border-color, #e0e1e9)}";
const ModusCardStyle0 = modusCardCss;

const ModusCard$1 = /*@__PURE__*/ proxyCustomElement(class ModusCard extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.ariaLabel = undefined;
        this.height = '269px';
        this.width = '240px';
        this.borderRadius = undefined;
        this.showCardBorder = true;
        this.showShadowOnHover = true;
    }
    render() {
        return (h("article", { key: '257c5bf3fe65862510aadcb6193acf8ac81ed87a', class: `${this.showShadowOnHover ? 'shadow' : ''} ${this.showCardBorder ? 'card-border' : ''}  `, "aria-label": this.ariaLabel || undefined, style: {
                height: this.height,
                width: this.width,
                'border-radius': this.borderRadius,
            } }, h("slot", { key: '945ca631265a2fd681feafd56ba8520c90a071e6' })));
    }
    static get style() { return ModusCardStyle0; }
}, [1, "modus-card", {
        "ariaLabel": [1, "aria-label"],
        "height": [1],
        "width": [1],
        "borderRadius": [1, "border-radius"],
        "showCardBorder": [4, "show-card-border"],
        "showShadowOnHover": [4, "show-shadow-on-hover"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-card"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-card":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusCard$1);
            }
            break;
    } });
}

const ModusCard = ModusCard$1;
const defineCustomElement = defineCustomElement$1;

export { ModusCard, defineCustomElement };
