// @trimble-oss/modus-web-components/dist/components/p-0ce126a9.js
import { p as proxyCustomElement, H, d as createEvent, h, c as Host } from './p-85681586.js';
import { I as IconClose } from './p-58bb9767.js';
import { d as defineCustomElement$1 } from './p-ba98e9ed.js';

const modusNavbarSearchOverlayCss = ":host{display:block}.overlay-wrapper{display:flex;flex-shrink:0;gap:45px}.overlay-wrapper .search-box{display:flex;flex-grow:2;justify-content:center}.overlay-wrapper .search-box modus-text-input{display:block;margin-left:1rem;max-width:764px;min-width:185px;width:100%}.overlay-wrapper .navbar-button{justify-content:end;padding-right:1rem}@media screen and (width <= 576px){.overlay-wrapper .navbar-button{padding-right:0.5rem}}";
const ModusNavbarSearchOverlayStyle0 = modusNavbarSearchOverlayCss;

const ModusNavbarSearchOverlay = /*@__PURE__*/ proxyCustomElement(class ModusNavbarSearchOverlay extends H {
    constructor() {
        super();
        this.__registerHost();
        this.close = createEvent(this, "close", 7);
        this.search = createEvent(this, "search", 7);
    }
    componentDidRender() {
        var _a;
        (_a = this.modusTextInput) === null || _a === void 0 ? void 0 : _a.focusInput();
    }
    render() {
        return (h(Host, { key: 'a46ddd8fae91428a953664b254022e34aec0faea' }, h("div", { key: 'a1a056390e6b0be61b8d1436b8a58e3edd4f1393', class: "overlay-wrapper" }, h("div", { key: '0e7e82a55e7f4d05e796e0aacb5d594fe9028641', class: "search-box" }, h("modus-text-input", { key: '03094e0fe493ecf14b192e59e5ab41a286301d13', placeholder: "Search", size: "large", type: "search", clearable: true, onValueChange: (event) => this.search.emit(event.detail), "include-search-icon": true, ref: (el) => (this.modusTextInput = el) })), h("div", { key: '1582aee27b56f2b681e86eca6ff2fd97d9637f5b', class: "navbar-button", "data-test-id": "close-button" }, h("span", { key: '97aac7cca2c843862fdf1a9d0339921f103faf59', class: "navbar-button-icon", role: "button", tabIndex: 0, onKeyDown: (event) => event.code === 'Enter' && this.close.emit(), onClick: () => this.close.emit() }, h(IconClose, { key: '2075c38346b3c9a853ff3657610ea9b2163801e4', size: "24" }))))));
    }
    get element() { return this; }
    static get style() { return ModusNavbarSearchOverlayStyle0; }
}, [0, "modus-navbar-search-overlay"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-navbar-search-overlay", "modus-text-input"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-navbar-search-overlay":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusNavbarSearchOverlay);
            }
            break;
        case "modus-text-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusNavbarSearchOverlay as M, defineCustomElement as d };
