// @trimble-oss/modus-web-components/dist/components/p-34f769d6.js
import { p as proxyCustomElement, H, d as createEvent, h } from './p-85681586.js';
import { I as IconCheck } from './p-c8e30a17.js';
import { M as ModusIconMap } from './p-cf76d4e3.js';

const modusListItemCss = "li{align-items:center;background-color:var(--modus-list-item-bg, #fff);border:1px solid var(--modus-list-item-border-color, #e0e1e9);border-radius:4px;box-sizing:border-box;color:var(--modus-list-item-color, #252a2e);display:flex;fill:var(--modus-list-item-color, #252a2e);font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";gap:1rem;justify-content:space-between;min-height:2.5rem;padding:0.375rem 0.75rem}li .text-container{display:flex;flex-direction:column;flex-grow:1;justify-content:center;min-width:0}li .slot{font-size:0.875rem;line-height:1.125rem;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}li .sub-text{display:block;font-size:0.75rem;line-height:1rem}li .wrap{word-wrap:break-word}li .no-wrap{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}li .icon-check{flex-shrink:0;margin-left:auto}li.borderless,li.borderless:hover:not(.disabled){border:none}li:hover:not(.disabled){background-color:var(--modus-list-item-hover-bg, #e0e1e9);border:1px solid var(--modus-list-item-border-color, #e0e1e9);color:var(--modus-list-item-color, #252a2e);fill:var(--modus-list-item-color, #252a2e)}li.small{gap:0.625rem;min-height:2rem;padding:0.25rem 0.625rem}li.small .slot{font-size:0.75rem;line-height:0.875rem}li.small .sub-text{font-size:0.625rem;line-height:0.75rem}li.large{min-height:3rem;padding:0.625rem 1rem}li.disabled{color:var(--modus-list-item-disabled-color, #b7b9c3);cursor:default;fill:var(--modus-list-item-disabled-color, #b7b9c3)}li.selected{background-color:var(--modus-list-item-selected-bg, #dcedf9);border:1px solid var(--modus-list-item-selected-border-color, #dcedf9);color:var(--modus-list-item-color, #252a2e);fill:var(--modus-list-item-color, #252a2e)}@media screen and (forced-colors: active){li.disabled{border-color:GrayText;color:GrayText}}";
const ModusListItemStyle0 = modusListItemCss;

const ModusListItem = /*@__PURE__*/ proxyCustomElement(class ModusListItem extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.itemClick = createEvent(this, "itemClick", 7);
        this.classBySize = new Map([
            ['condensed', 'small'],
            ['standard', 'standard'],
            ['large', 'large'],
        ]);
        this.borderless = undefined;
        this.disabled = undefined;
        this.selected = undefined;
        this.leftIcon = undefined;
        this.size = 'standard';
        this.subText = undefined;
        this.wrapSubText = true;
        this.type = 'standard';
        this.iconColor = undefined;
    }
    async focusItem() {
        var _a;
        (_a = this.listItemRef) === null || _a === void 0 ? void 0 : _a.focus();
    }
    handleKeydown(e) {
        if (e.key.toLowerCase() === 'enter' && !this.disabled) {
            this.itemClick.emit();
        }
    }
    render() {
        const containerClass = `${this.classBySize.get(this.size)} ${this.disabled ? 'disabled' : ''} ${this.selected ? 'selected' : ''} ${this.borderless ? 'borderless' : ''}`;
        const iconCheckSize = this.size === 'condensed' ? '16' : '24';
        return (h("li", { key: 'a143f4008c7667ab3a118f8e4f243fdf845c9a25', ref: (el) => (this.listItemRef = el), part: "list-item-li", class: containerClass, tabIndex: this.disabled ? -1 : 0, onClick: () => (!this.disabled ? this.itemClick.emit() : null), onKeyDown: (e) => this.handleKeydown(e) }, this.leftIcon && (h("span", { class: "icon left-icon" }, h(ModusIconMap, { icon: this.leftIcon, size: "24", color: this.iconColor }))), h("div", { key: '6caeadbfe730b540ff4b34d3c1991c5c64a95add', class: "text-container" }, h("span", { key: 'd0d695a03abafd70b5a240178921492c0523e58c', class: "slot" }, h("slot", { key: 'f3f949fbe697d45e4ae93429de1c6e753354ff39' })), this.subText && h("span", { class: 'sub-text ' + (this.wrapSubText ? 'wrap' : 'no-wrap') }, this.subText)), this.selected && h(IconCheck, { size: iconCheckSize })));
    }
    static get style() { return ModusListItemStyle0; }
}, [1, "modus-list-item", {
        "borderless": [4],
        "disabled": [4],
        "selected": [4],
        "leftIcon": [1, "left-icon"],
        "size": [1],
        "subText": [1, "sub-text"],
        "wrapSubText": [4, "wrap-sub-text"],
        "type": [1],
        "iconColor": [1, "icon-color"],
        "focusItem": [64]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-list-item"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-list-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusListItem);
            }
            break;
    } });
}

export { ModusListItem as M, defineCustomElement as d };
