// @trimble-oss/modus-web-components/dist/components/p-5d1d55be.js
const HTML_ATTR_DATA_ACCESSOR_KEY = 'data-accessor-key';
const COLUMN_DEF_DATATYPE_KEY = 'dataType';
const COLUMN_DEF_SUB_ROWS_KEY = 'subRows';
const COLUMN_DEF_CELL_EDITOR_TYPE_KEY = 'cellEditorType';
const COLUMN_DEF_CELL_EDITOR_ARGS_KEY = 'cellEditorArgs';
const COLUMN_DEF_SHOWTOTAL = 'showTotal';
const COLUMN_DEF_ROW_SELECTION_ID = 'row-checkbox';
const COLUMN_DEF_ROW_SELECTION_CSS = 'row-checkbox';
const COLUMN_DEF_DATATYPE_CUSTOM = 'custom';
const COLUMN_DEF_DATATYPE_TEXT = 'text';
const COLUMN_DEF_DATATYPE_INTEGER = 'integer';
const COLUMN_DEF_DATATYPE_LINK = 'link';
const COLUMN_DEF_DATATYPE_DATE = 'date';
const COLUMN_DEF_DATATYPE_BADGE = 'badge';
const CELL_EDIT_TYPE_TEXT = COLUMN_DEF_DATATYPE_TEXT;
const CELL_EDIT_TYPE_INT = COLUMN_DEF_DATATYPE_INTEGER;
const CELL_EDIT_TYPE_SELECT = 'select';
const CELL_EDIT_TYPE_AUTOCOMPLETE = 'autocomplete';
const CELL_EDIT_TYPE_DATE = COLUMN_DEF_DATATYPE_DATE;
const ALLOWED_CELL_EDIT_TYPES = [
    CELL_EDIT_TYPE_TEXT,
    CELL_EDIT_TYPE_INT,
    CELL_EDIT_TYPE_SELECT,
    CELL_EDIT_TYPE_AUTOCOMPLETE,
    CELL_EDIT_TYPE_DATE,
];
const PAGINATION_DEFAULT_SIZES = [10, 20, 30];
// Keyboard key constants
const KEYBOARD_ENTER = 'enter';
const KEYBOARD_TAB = 'tab';
const KEYBOARD_LEFT = 'arrowleft';
const KEYBOARD_RIGHT = 'arrowright';
const KEYBOARD_UP = 'arrowup';
const KEYBOARD_DOWN = 'arrowdown';
const KEYBOARD_SPACE = ' ';
const KEYBOARD_ESCAPE = 'escape';
const EXPANDED_STATE_KEY = 'expanded';
const SORTING_STATE_KEY = 'sorting';
const ROW_SELECTION_STATE_KEY = 'rowSelection';
const COLUMN_SIZING_INFO_STATE_KEY = 'columnSizingInfo';
const PAGINATION_STATE_KEY = 'pagination';
const COLUMN_SIZING_STATE_KEY = 'columnSizing';
const COLUMN_VISIBILITY_STATE_KEY = 'columnVisibility';
const COLUMN_ORDER_STATE_KEY = 'columnOrder';

export { ALLOWED_CELL_EDIT_TYPES as A, COLUMN_DEF_CELL_EDITOR_TYPE_KEY as B, COLUMN_DEF_ROW_SELECTION_ID as C, EXPANDED_STATE_KEY as E, HTML_ATTR_DATA_ACCESSOR_KEY as H, KEYBOARD_RIGHT as K, PAGINATION_DEFAULT_SIZES as P, ROW_SELECTION_STATE_KEY as R, SORTING_STATE_KEY as S, COLUMN_DEF_ROW_SELECTION_CSS as a, COLUMN_DEF_DATATYPE_KEY as b, COLUMN_DEF_DATATYPE_INTEGER as c, COLUMN_DEF_SHOWTOTAL as d, KEYBOARD_LEFT as e, KEYBOARD_ENTER as f, KEYBOARD_ESCAPE as g, KEYBOARD_TAB as h, COLUMN_DEF_SUB_ROWS_KEY as i, KEYBOARD_SPACE as j, PAGINATION_STATE_KEY as k, COLUMN_SIZING_STATE_KEY as l, COLUMN_VISIBILITY_STATE_KEY as m, COLUMN_ORDER_STATE_KEY as n, COLUMN_SIZING_INFO_STATE_KEY as o, CELL_EDIT_TYPE_TEXT as p, CELL_EDIT_TYPE_INT as q, CELL_EDIT_TYPE_DATE as r, CELL_EDIT_TYPE_AUTOCOMPLETE as s, CELL_EDIT_TYPE_SELECT as t, KEYBOARD_UP as u, KEYBOARD_DOWN as v, COLUMN_DEF_CELL_EDITOR_ARGS_KEY as w, COLUMN_DEF_DATATYPE_BADGE as x, COLUMN_DEF_DATATYPE_CUSTOM as y, COLUMN_DEF_DATATYPE_LINK as z };
