// @trimble-oss/modus-web-components/dist/components/p-6abe744a.js
import { p as proxyCustomElement, H, d as createEvent, h as h$1, c as Host } from './p-85681586.js';
import { f as KEYBOARD_ENTER, j as KEYBOARD_SPACE, g as KEYBOARD_ESCAPE, w as COLUMN_DEF_CELL_EDITOR_ARGS_KEY, b as COLUMN_DEF_DATATYPE_KEY, p as CELL_EDIT_TYPE_TEXT, A as ALLOWED_CELL_EDIT_TYPES, x as COLUMN_DEF_DATATYPE_BADGE, c as COLUMN_DEF_DATATYPE_INTEGER, y as COLUMN_DEF_DATATYPE_CUSTOM, z as COLUMN_DEF_DATATYPE_LINK, B as COLUMN_DEF_CELL_EDITOR_TYPE_KEY } from './p-5d1d55be.js';
import { N as NavigateTableCells } from './p-3af926db.js';
import { h } from './p-5f965237.js';
import { d as defineCustomElement$b } from './p-378ceba8.js';
import { d as defineCustomElement$a } from './p-57104bdf.js';
import { d as defineCustomElement$9 } from './p-5d4f663e.js';
import { d as defineCustomElement$8 } from './p-712cbdfb.js';
import { d as defineCustomElement$7 } from './p-854b47fa.js';
import { d as defineCustomElement$6 } from './p-f5c6b288.js';
import { d as defineCustomElement$5 } from './p-92734348.js';
import { d as defineCustomElement$4 } from './p-9158ea44.js';
import { d as defineCustomElement$3 } from './p-1b7ef7fc.js';
import { d as defineCustomElement$2 } from './p-25e7c862.js';
import { d as defineCustomElement$1 } from './p-ba98e9ed.js';
import { c as createPopper } from './p-68a5aaff.js';

// Code taken from - https://github.com/TanStack/table/blob/main/packages/react-table/src/index.tsx
/**
 * Cell formatter enables formatting of a cell value.
 */
function CellFormatter(Comp, props) {
    return !Comp ? null : isFunction(Comp) ? h(Comp, Object.assign({}, props)) : Comp;
}
/**
 * Checks if the input for cell formatter is a function object.
 */
function isFunction(component) {
    return isClassComponent(component) || typeof component === 'function';
}
function isClassComponent(component) {
    return (typeof component === 'function' &&
        (() => {
            const proto = Object.getPrototypeOf(component);
            return proto.prototype && proto.prototype.isFunctionalComponent;
        })());
}

const ModusTableCellLinkElement = ({ link, onLinkClick }) => {
    function handleLinkKeyDown(e) {
        const key = e.key.toLowerCase();
        if (key === KEYBOARD_ENTER || key === KEYBOARD_SPACE) {
            onLinkClick(link);
            e.stopImmediatePropagation();
        }
    }
    return (h("div", { class: "cell-link", tabIndex: 0, onClick: () => onLinkClick(link), onKeyDown: handleLinkKeyDown }, link.display));
};

const ModusTableCellBadgeElement = ({ badge, onBadgeClick }) => {
    function handleBadgeKeyDown(e) {
        const key = e.key.toLowerCase();
        if (key === KEYBOARD_ENTER || key === KEYBOARD_SPACE) {
            onBadgeClick(badge);
            e.stopImmediatePropagation();
        }
    }
    return (h("div", { class: "cell-badge", onClick: () => onBadgeClick(badge), onKeyDown: () => handleBadgeKeyDown },
        h("modus-badge", { tabIndex: 0, color: badge.color, size: badge.size, type: badge.type, ariaLabel: badge === null || badge === void 0 ? void 0 : badge.ariaLabel }, badge.text)));
};

// eslint-disable-next-line
const IconChevronDownThick = (props) => {
    var _a, _b, _c;
    return (h("svg", { class: `icon-chevron-down-thick ${props.pressed ? 'pressed' : ''}`, fill: (_a = props.color) !== null && _a !== void 0 ? _a : 'currentColor', height: (_b = props.size) !== null && _b !== void 0 ? _b : 16, width: (_c = props.size) !== null && _c !== void 0 ? _c : 16, onClick: props.onClick, viewBox: "0 0 24 24" },
        h("path", { d: "M15.88 9.29 12 13.17 8.12 9.29a.996.996 0 1 0-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z" })));
};

// eslint-disable-next-line
const IconChevronUpThick = (props) => {
    var _a, _b, _c;
    return (h("svg", { class: `icon-chevron-up-thick ${props.pressed ? 'pressed' : ''}`, fill: (_a = props.color) !== null && _a !== void 0 ? _a : 'currentColor', height: (_b = props.size) !== null && _b !== void 0 ? _b : 16, width: (_c = props.size) !== null && _c !== void 0 ? _c : 16, onClick: props.onClick, viewBox: "0 0 24 24" },
        h("path", { d: "M11.29 8.71 6.7 13.3a.996.996 0 1 0 1.41 1.41L12 10.83l3.88 3.88a.996.996 0 1 0 1.41-1.41L12.7 8.71a.996.996 0 0 0-1.41 0z" })));
};

const ModusTableCellExpandIcons = ({ row }) => {
    let expandEl;
    return (h("span", { class: "expand-icon-container", ref: (ref) => (expandEl = ref), style: { paddingLeft: `${row.depth * 2}rem` }, onClick: (e) => {
            row.getToggleExpandedHandler()();
            e.stopImmediatePropagation();
        } }, row.getCanExpand() && (h("span", { class: "expand-icon", tabIndex: 0, onKeyDown: (event) => {
            if (event.key.toLowerCase() === KEYBOARD_ENTER || event.key.toLowerCase() === KEYBOARD_SPACE) {
                expandEl.click();
                event.stopImmediatePropagation();
            }
        } }, row.getIsExpanded() ? h(IconChevronUpThick, { size: '24' }) : h(IconChevronDownThick, { size: '24' })))));
};

const ModusTableCustomCellElement = ({ customElement }) => {
    return h("div", { class: "cell-custom" },
        " ",
        h("flex-renderer", { content: customElement.getValue().toString() }));
};

const ModusTableCellMain = /*@__PURE__*/ proxyCustomElement(class ModusTableCellMain extends H {
    constructor() {
        super();
        this.__registerHost();
        this.cellInputValueChange = createEvent(this, "cellInputValueChange", 7);
        this.onCellClick = (e) => this.handleCellClick(e);
        this.onCellKeyDown = (e) => this.handleCellKeydown(e);
        this.onCellBlur = (e) => this.handleCellBlur(e);
        this.cellEditableKey = 'cellEditable';
        this.accessorKey = 'accessorKey';
        this.isCellEditable = () => {
            var _a;
            const editable = this.cell.column.columnDef[this.cellEditableKey];
            const isDisabledFn = (_a = this.getEditorArgs()) === null || _a === void 0 ? void 0 : _a.isCellDisabled;
            const isDisabled = typeof isDisabledFn === 'function' ? isDisabledFn(this.cell.row.original) : false;
            return editable && !isDisabled;
        };
        this.handleCellClick = (event) => {
            if (event.defaultPrevented)
                return;
            if (this.isCellEditable()) {
                this.editMode = true;
            }
        };
        this.handleCellBlur = (event) => {
            if (!this.el.contains(event.relatedTarget)) {
                this.editMode = false;
            }
        };
        this.handleCellKeydown = (event) => {
            if (event.defaultPrevented)
                return;
            if (this.isCellEditable()) {
                this.editMode = true;
                event.stopPropagation();
            }
            else {
                NavigateTableCells({
                    eventKey: event.key,
                    cellElement: this.cellEl,
                });
            }
        };
        this.handleCellEditorOnInputChange = (newValue, oldValue) => {
            this.cellInputValueChange.emit({
                row: this.cell.row,
                accessorKey: this.cell.column.columnDef[this.accessorKey],
                newValue,
                oldValue,
            });
        };
        this.handleCellEditorKeyDown = (event, newValue, oldValue) => {
            var _a;
            const key = (_a = event.key) === null || _a === void 0 ? void 0 : _a.toLowerCase();
            if (key === KEYBOARD_ENTER) {
                this.handleCellEditorValueChange(newValue, oldValue);
                NavigateTableCells({
                    eventKey: KEYBOARD_ENTER,
                    cellElement: this.cellEl,
                });
            }
            else if (key === KEYBOARD_ESCAPE) {
                this.editMode = false;
                this.cellEl.focus();
                this.destroyErrorTooltip();
            }
            else
                return;
            event.stopPropagation();
        };
        this.cell = undefined;
        this.context = undefined;
        this.hasRowsExpandable = undefined;
        this.valueChange = undefined;
        this.errorMessage = undefined;
        this.editMode = undefined;
    }
    onEditModeChange(newValue) {
        if (newValue) {
            this.cellEl.classList.add('edit-mode');
            this.createErrorTooltip(); // Create tooltip when entering edit mode
            if (this.errorMessage)
                this.showErrorTooltip();
        }
        else {
            this.cellEl.classList.remove('edit-mode');
            this.destroyErrorTooltip(); // Destroy tooltip when exiting edit mode
        }
    }
    onContextChange() {
        this.updateErrorState();
    }
    connectedCallback() {
        this.cellEl = this.el.parentElement;
        this.cellEl.addEventListener('click', this.onCellClick);
        this.cellEl.addEventListener('keydown', this.onCellKeyDown);
        this.cellEl.addEventListener('blur', this.onCellBlur);
        this.updateErrorState();
    }
    disconnectedCallback() {
        if (this.cellEl) {
            this.cellEl.removeEventListener('click', this.onCellClick);
            this.cellEl.removeEventListener('keydown', this.onCellKeyDown);
            this.cellEl.removeEventListener('blur', this.onCellBlur);
        }
        this.destroyErrorTooltip();
    }
    updateErrorState() {
        var _a, _b, _c, _d, _e;
        const rowId = (_a = this.cell.row.id) !== null && _a !== void 0 ? _a : this.cell.row.index;
        const accessorKey = this.cell.column.columnDef[this.accessorKey];
        const errorMessage = (_c = (_b = this.context.errors) === null || _b === void 0 ? void 0 : _b[rowId]) === null || _c === void 0 ? void 0 : _c[accessorKey];
        if (errorMessage) {
            this.errorMessage = errorMessage;
            (_d = this.cellEl) === null || _d === void 0 ? void 0 : _d.classList.add('error');
            this.showErrorTooltip();
        }
        else {
            this.errorMessage = '';
            (_e = this.cellEl) === null || _e === void 0 ? void 0 : _e.classList.remove('error');
            this.hideErrorTooltip();
        }
    }
    getEditorType() {
        const editorType = this.cell.column.columnDef[COLUMN_DEF_CELL_EDITOR_TYPE_KEY];
        const dataType = this.cell.column.columnDef[COLUMN_DEF_DATATYPE_KEY];
        let editorTypeToReturn = CELL_EDIT_TYPE_TEXT;
        if (!editorType) {
            if (ALLOWED_CELL_EDIT_TYPES.includes(dataType))
                editorTypeToReturn = dataType;
        }
        else {
            editorTypeToReturn = editorType;
        }
        return editorTypeToReturn;
    }
    getEditorArgs() {
        return this.cell.column.columnDef[COLUMN_DEF_CELL_EDITOR_ARGS_KEY];
    }
    /**
     * Returns whether a cell is editable based on row index and column ID.
     * @param rowIndex The index of the row.
     * @param columnId The ID of the column.
     * @returns Boolean indicating if the cell is editable.
     */
    async handleCellEdit(rowIndex, columnId) {
        const tableInstance = this.cell.getContext().table;
        const row = tableInstance.getRowModel().rows[rowIndex];
        if (!row)
            return;
        const cell = row.getAllCells().find((cell) => cell.column.id === columnId);
        if (!cell)
            return;
        // Focus on the cell element
        const cellElement = this.el.querySelector(`[data-cell-id="${rowIndex}-${columnId}"]`);
        if (cellElement) {
            cellElement.focus();
        }
        this.editMode = true;
    }
    handleCellEditorValueChange(newValue, oldValue) {
        if (this.editMode && newValue !== oldValue && this.valueChange) {
            this.valueChange({
                row: this.cell.row,
                accessorKey: this.cell.column.columnDef[this.accessorKey],
                newValue,
                oldValue,
            });
        }
        this.editMode = false;
    }
    renderCellValue() {
        var _a;
        const { row, getValue } = this.cell;
        const cellValue = getValue();
        const cell = this.cell;
        if (cellValue === null || cellValue === undefined)
            return null;
        const { cellLinkClick, wrapText } = this.context;
        const cellDataType = (_a = cellValue['_type']) !== null && _a !== void 0 ? _a : this.cell.column.columnDef[COLUMN_DEF_DATATYPE_KEY];
        const wrap = cellDataType === COLUMN_DEF_DATATYPE_BADGE ? false : wrapText;
        const classes = {
            'cell-content': true,
            'truncate-text': !wrap,
            'wrap-text': wrap,
            'text-align-right': cellDataType === COLUMN_DEF_DATATYPE_INTEGER,
        };
        const renderCell = () => {
            if (cellDataType === COLUMN_DEF_DATATYPE_CUSTOM) {
                return h$1(ModusTableCustomCellElement, { customElement: cell });
            }
            else if (cellDataType === COLUMN_DEF_DATATYPE_LINK) {
                return (h$1(ModusTableCellLinkElement, { link: cellValue, onLinkClick: (link) => {
                        this.cellEl.focus();
                        cellLinkClick.emit(link);
                    } }));
            }
            else if (cellDataType === COLUMN_DEF_DATATYPE_BADGE) {
                return (h$1(ModusTableCellBadgeElement, { badge: cellValue, onBadgeClick: () => {
                        this.cellEl.focus();
                    } }));
            }
            else {
                return CellFormatter(this.cell.column.columnDef.cell, this.cell.getContext());
            }
        };
        return (h$1("div", { class: classes }, this.hasRowsExpandable && h$1(ModusTableCellExpandIcons, { row: row }), h$1("span", { class: wrap ? 'wrap-text' : 'truncate-text' }, renderCell())));
    }
    createErrorTooltip() {
        if (!this.errorTooltip) {
            this.errorTooltip = document.createElement('div');
            this.errorTooltip.className = 'error-tooltip';
            this.cellEl.appendChild(this.errorTooltip);
            this.popperInstance = createPopper(this.cellEl, this.errorTooltip, {
                placement: 'bottom-start',
                modifiers: [
                    {
                        name: 'offset',
                        options: {
                            offset: [0.2, 0.2], // Offset from the element
                            mainAxis: false,
                        },
                    },
                    {
                        name: 'preventOverflow',
                        options: {
                            boundary: 'viewport',
                        },
                    },
                ],
            });
        }
    }
    showErrorTooltip() {
        if (this.errorTooltip) {
            this.errorTooltip.innerText = 'Invalid Input';
            this.errorTooltip.style.display = 'block';
            if (this.popperInstance) {
                this.popperInstance.update();
            }
        }
    }
    hideErrorTooltip() {
        if (this.errorTooltip) {
            this.errorTooltip.style.display = 'none';
        }
    }
    destroyErrorTooltip() {
        if (this.popperInstance) {
            this.popperInstance.destroy();
            this.popperInstance = null;
        }
        if (this.errorTooltip) {
            this.errorTooltip.remove();
            this.errorTooltip = null;
        }
    }
    render() {
        var _a;
        const valueString = (_a = this.cell.getValue()) === null || _a === void 0 ? void 0 : _a.toString();
        return (h$1(Host, { key: 'f83e4a3516e7e09c5ac025e612e16444e9a124dc' }, this.editMode ? (h$1("modus-table-cell-editor", { "data-type": this.cell.column.columnDef[COLUMN_DEF_DATATYPE_KEY], value: this.cell.getValue(), type: this.getEditorType(), args: this.getEditorArgs(), valueChange: (newVal) => this.handleCellEditorValueChange(newVal, valueString), keyDown: (event, newVal) => this.handleCellEditorKeyDown(event, newVal, valueString), inputValueChangeHandler: (newVal) => this.handleCellEditorOnInputChange(newVal, valueString) })) : (this.renderCellValue())));
    }
    get el() { return this; }
    static get watchers() { return {
        "editMode": ["onEditModeChange"],
        "context": ["onContextChange"]
    }; }
}, [0, "modus-table-cell-main", {
        "cell": [16],
        "context": [16],
        "hasRowsExpandable": [4, "has-rows-expandable"],
        "valueChange": [16],
        "errorMessage": [32],
        "editMode": [32],
        "handleCellEdit": [64]
    }, undefined, {
        "editMode": ["onEditModeChange"],
        "context": ["onContextChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-cell-main", "flex-renderer", "modus-autocomplete", "modus-badge", "modus-chip", "modus-date-input", "modus-date-picker", "modus-number-input", "modus-select", "modus-spinner", "modus-table-cell-editor", "modus-text-input"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-cell-main":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTableCellMain);
            }
            break;
        case "flex-renderer":
            if (!customElements.get(tagName)) {
                defineCustomElement$b();
            }
            break;
        case "modus-autocomplete":
            if (!customElements.get(tagName)) {
                defineCustomElement$a();
            }
            break;
        case "modus-badge":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "modus-chip":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "modus-date-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "modus-date-picker":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "modus-number-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "modus-select":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "modus-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-table-cell-editor":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-text-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusTableCellMain as M, defineCustomElement as d };
