// @trimble-oss/modus-web-components/dist/components/p-8f144f06.js
import { p as proxyCustomElement, H, h, c as Host } from './p-85681586.js';
import { d as defineCustomElement$4 } from './p-56367692.js';
import { d as defineCustomElement$3 } from './p-a67f7d10.js';
import { d as defineCustomElement$2 } from './p-97d71c36.js';
import { d as defineCustomElement$1 } from './p-affe4060.js';

const modusTableToolbarCss = ".table-toolbar{align-items:center;display:flex;justify-content:space-between;padding-bottom:3px}.table-toolbar .section{align-items:center;display:flex}";
const ModusTableToolbarStyle0 = modusTableToolbarCss;

const ModusTablePanel = /*@__PURE__*/ proxyCustomElement(class ModusTablePanel extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.context = undefined;
    }
    render() {
        // const { tableInstance: table, toolbarOptions: options } = this.context;
        return (h(Host, { key: 'd8b9a5275dc4e77bdacb4958d1cdd4079872302c' }, h("div", { key: '5f5d87db6a9cd8dd774e055396fd12679d8c534e', class: "table-toolbar" }, h("div", { key: '432d5965c73c369d63e55f81ce300dc9c52caee6', class: "section" }, h("slot", { key: '11786f71e2d0a2ac80bbf56a5e3de148a57de865', name: "group-left" })), h("div", { key: '2b515ed9f8869b8c16ef478deb87fb5f629182cf', class: "section" }, h("slot", { key: 'ea1db6297b8534aa58ed9c5a5793b810ff7b8dc3', name: "group-right" }), h("modus-table-dropdown-menu", { context: this.context })))));
    }
    static get style() { return ModusTableToolbarStyle0; }
}, [1, "modus-table-toolbar", {
        "context": [16]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-toolbar", "modus-button", "modus-checkbox", "modus-table-columns-visibility", "modus-table-dropdown-menu"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-toolbar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTablePanel);
            }
            break;
        case "modus-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "modus-checkbox":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-table-columns-visibility":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-table-dropdown-menu":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusTablePanel as M, defineCustomElement as d };
