// @trimble-oss/modus-web-components/dist/components/p-aec13d9d.js
import { p as proxyCustomElement, H, d as createEvent, h } from './p-85681586.js';
import { M as ModusIconMap } from './p-cf76d4e3.js';
import { d as defineCustomElement$2 } from './p-dca65392.js';
import { d as defineCustomElement$1 } from './p-34f769d6.js';

const modusNavbarProfileMenuCss = ".profile-menu{background-color:var(--modus-body-bg, #fff);border:0.0625rem solid var(--modus-navbar-popup-border-color, #cbcdd6);border-radius:0.25rem;box-shadow:0 0 8px rgba(36, 35, 45, 0.3);color:var(--modus-body-color, #252a2e);display:flex;flex-direction:column;max-width:384px;padding:0;position:absolute;right:0;top:48px;z-index:100;align-items:center;background-color:var(--modus-navbar-popup-bg, #fff);cursor:default;padding:0;width:19rem}.profile-menu .user{align-items:center;display:flex;flex-direction:column;justify-content:center;padding:1rem 1.5rem 0 1.5rem}.profile-menu .user .avatar,.profile-menu .user .initials{background-color:transparent;border-radius:50%;height:6rem;margin-bottom:1rem;user-select:none;width:6rem}.profile-menu .user .initials{align-items:center;background-color:var(--modus-navbar-icon-color, #252a2e);color:var(--modus-navbar-profile-icon-initials-color, #fff);display:flex;font-size:1.5rem;font-weight:600;justify-content:center}.profile-menu .user .user-details{align-items:center;display:flex;flex-direction:column;justify-content:center;margin-bottom:1rem;max-width:16rem}.profile-menu .user .user-details .username{font-size:1rem;font-weight:600;overflow-wrap:anywhere;text-align:center;width:fit-content}.profile-menu .user .user-details .email{font-size:0.75rem;overflow-wrap:anywhere;text-align:center}.profile-menu.profile-menu-blue .user .avatar,.profile-menu.profile-menu-blue .user .initials{background-color:#b7b9c3;color:var(--modus-navbar-blue-profile-icon-initials-color, #0e416c)}.profile-menu .links{cursor:pointer;margin-bottom:0.75rem;max-width:19rem;user-select:none;width:100%}.profile-menu .links .link-item{align-items:center;display:flex;flex-direction:row;white-space:normal;word-wrap:break-word}.profile-menu .links .icon{align-items:center;display:flex;justify-content:center;margin-right:1rem}.profile-menu .sign-out{background-color:var(--modus-navbar-profile-menu-sign-out-background-color, #f1f1f6);user-select:none;width:100%}.profile-menu .sign-out div{color:var(--modus-blue, #0063a3);cursor:pointer;font-size:0.875rem;padding:0.625rem 1rem}@media (hover: hover){.profile-menu .sign-out div:hover{text-decoration:underline}}.profile-menu.reverse{left:0}";
const ModusNavbarProfileMenuStyle0 = modusNavbarProfileMenuCss;

const ModusNavbarProfileMenu = /*@__PURE__*/ proxyCustomElement(class ModusNavbarProfileMenu extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.linkClick = createEvent(this, "linkClick", 7);
        this.signOutClick = createEvent(this, "signOutClick", 7);
        this.avatarUrl = undefined;
        this.email = undefined;
        this.initials = undefined;
        this.links = undefined;
        this.signOutText = 'Sign out';
        this.reverse = undefined;
        this.username = undefined;
        this.variant = 'default';
    }
    signOutKeydownHandler(event) {
        if (event.code !== 'Enter') {
            return;
        }
        this.signOutClick.emit();
    }
    render() {
        var _a;
        const direction = this.reverse ? 'reverse' : '';
        const variant = this.variant === 'default' ? '' : `profile-menu-${this.variant}`;
        return (h("div", { key: '90cca497d06325e8ff3b0bb776ff7ee115fc2d85', class: `profile-menu ${direction} ${variant}`, onClick: (event) => event.preventDefault() }, h("div", { key: '1b6d24352bd6e5188b96c585115f88a142099fba', class: "user" }, this.avatarUrl ? (h("img", { class: "avatar", src: this.avatarUrl, alt: "Avatar" })) : (h("span", { class: "initials" }, this.initials)), h("div", { key: '6a9504936beef21a16c5fc730f818cca46546660', class: "user-details" }, h("div", { key: '993b31653b5104f56dd0a494fc7df75683640fe0', class: "username" }, this.username), h("div", { key: '92b8228a67399a964daf78ed4e2e6536498f5d35', class: "email" }, this.email))), ((_a = this.links) === null || _a === void 0 ? void 0 : _a.length) ? (h("modus-list", { class: "links" }, this.links.map((link) => {
            return (h("modus-list-item", { borderless: true, onItemClick: () => this.linkClick.emit(link.id) }, h("div", { class: "link-item" }, link.icon && (h("div", { class: "icon" }, h(ModusIconMap, { icon: link.icon, size: "24" }))), link.display)));
        }))) : null, h("slot", { key: 'f30f7d8f565805b4a500f0fa3d91a99f303b6e59' }), h("div", { key: '1697274a71163193dec204390aefd34ed3762fb5', class: "sign-out", onClick: () => this.signOutClick.emit(), onKeyDown: (event) => this.signOutKeydownHandler(event), tabIndex: 0 }, h("div", { key: '4497874cf435a21686d9415c96a2ecb7dd344c66' }, this.signOutText))));
    }
    static get style() { return ModusNavbarProfileMenuStyle0; }
}, [1, "modus-navbar-profile-menu", {
        "avatarUrl": [1, "avatar-url"],
        "email": [1],
        "initials": [1],
        "links": [16],
        "signOutText": [1, "sign-out-text"],
        "reverse": [4],
        "username": [1],
        "variant": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-navbar-profile-menu", "modus-list", "modus-list-item"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-navbar-profile-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusNavbarProfileMenu);
            }
            break;
        case "modus-list":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-list-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusNavbarProfileMenu as M, defineCustomElement as d };
