// @trimble-oss/modus-web-components/dist/components/p-ce52cff9.js
import { p as proxyCustomElement, H, h } from './p-85681586.js';

const modusDividerCss = ".modus-divider{border-left:1px solid var(--modus-toolbar-divider-bg, #252a2e);height:32px;margin-left:3px;margin-right:3px;opacity:0.25}";
const ModusDividerStyle0 = modusDividerCss;

const ModusDivider = /*@__PURE__*/ proxyCustomElement(class ModusDivider extends H {
    constructor() {
        super();
        this.__registerHost();
    }
    render() {
        return h("div", { key: 'fff79c4a9eb431f2178f2b29f5bdb34f48109b27', class: "modus-divider" });
    }
    static get style() { return ModusDividerStyle0; }
}, [0, "modus-divider"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-divider"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-divider":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusDivider);
            }
            break;
    } });
}

export { ModusDivider as M, defineCustomElement as d };
