// @trimble-oss/modus-web-components/dist/components/p-dca65392.js
import { p as proxyCustomElement, H, h } from './p-85681586.js';

const modusListCss = "ul{background-color:var(--modus-list-bg, #fff);border-radius:0.25rem;display:flex;flex-direction:column;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";list-style:none;margin:0;padding:0;position:relative}";
const ModusListStyle0 = modusListCss;

const ModusList = /*@__PURE__*/ proxyCustomElement(class ModusList extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
    }
    handleKeyDown(e) {
        const itemsLength = this.element.children.length;
        if (e.key.toLowerCase() === 'arrowdown') {
            const index = Array.prototype.indexOf.call(this.element.children, e.target);
            let next = this.element.children.item((index + 1) % itemsLength);
            while (next === null || next === void 0 ? void 0 : next.disabled) {
                next = this.element.children.item((index + 2) % itemsLength);
            }
            next === null || next === void 0 ? void 0 : next.focusItem();
            e.preventDefault();
        }
        if (e.key.toLowerCase() === 'arrowup') {
            const index = Array.prototype.indexOf.call(this.element.children, e.target);
            let prev = this.element.children.item((index - 1) % itemsLength);
            while (prev === null || prev === void 0 ? void 0 : prev.disabled) {
                prev = this.element.children.item((index - 2) % itemsLength);
            }
            prev === null || prev === void 0 ? void 0 : prev.focusItem();
            e.preventDefault();
        }
    }
    render() {
        return (h("ul", { key: 'd50f4a349d16e3e76bca4e5923ffafb6580d9a80', part: "list-items", onKeyDown: (e) => this.handleKeyDown(e) }, h("slot", { key: '61d4f7de300ae52f88732311aa5c0b81ae3acbb8' })));
    }
    get element() { return this; }
    static get style() { return ModusListStyle0; }
}, [1, "modus-list"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-list"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-list":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusList);
            }
            break;
    } });
}

export { ModusList as M, defineCustomElement as d };
